xclim/utils.py

```diff
@@ -124,6 +124,143 @@ def percentile_doy(arr, window=5, per=.1):
+    else:
+        lat_cond = (da.lat >= da.lat.min()) & (da.lat <= da.lat.max())
+
+    if year_bnds:
+        year_bnds = np.asarray(year_bnds)
+        if len(year_bnds) == 1:
+            time_cond = da.time.dt.year == year_bnds
+        else:
+            time_cond = (da.time.dt.year >= year_bnds.min()) & (da.time.dt.year <= year_bnds.max())
+    else:
+        time_cond = (da.time.dt.year >= da.time.dt.year.min()) & (da.time.dt.year <= da.time.dt.year.max())
+
+    return da.where(lon_cond & lat_cond & time_cond, drop=True)
+
+
+def subset_gridpoint(da, lon, lat, year_bnds=''):
+    """Extract a nearest gridpoint from datarray based on lat lon coordinate.
+    Time series can optionally be subsetted by year(s)
+
+            Return a subsetted data array (or dataset) for the grid point falling nearest the input longitude and latitude
+            coordinates. Optionally subset the data array for years falling within provided year bounds
+
+            Parameters
+            ----------
+            da : xarray.DataArray or xarray.DataSet
+              Input data.
+            lon : float
+              longitude coordinate.
+            lat:  float
+              latitude coordinate.
+            year_bnds (optional) :  list of floats or float
+              List of maximum and minimum year bounds. If single float returns one year only.
+              Defaults to all years in original data-array
+
+
+            Returns
+            -------
+            xarray.DataArray or xarray.DataSet
+              subsetted data array or dataset
+
+        Examples
+        --------
+
+        >>> from xclim import utils
+        >>> ds = xr.open_dataset('pr.day.nc')
+        Subset lat lon point and multiple years
+        >>> prSub = utils.subset_gridpoint(ds.pr, lon=-75,lat=45,year_bnds=[1990,1999])
+        Subset lat, lon point and single year
+        >>> prSub = utils.subset_gridpoint(ds.pr, lon=-75,lat=45,year_bnds=1990) # one year only
+         Subset mutliple variables in a single dataset
+        >>> ds = xr.open_mfdataset(['pr.day.nc','tas.day.nc'])
+        >>> dsSub = utils.subset_gridpoint(ds, lon=-75,lat=45,year_bnds=[1990,1999])
+            """
+
+    # adjust for files with all postive longitudes if necessary
+    if np.all(da.lon > 0) and lon < 0:
+        lon += 360
+
+    try:
+        out = da.sel(lon=lon, lat=lat, method='nearest')
+
+    except:
+
+        dist = np.hypot(da.lat - lat, da.lon - lon)
+        iy, ix = np.unravel_index(np.argmin(dist, axis=None), da.lat.shape)
+        xydims = [x for x in da.dims if 'time' not in x]
+        args = {}
+        args[xydims[0]] = iy
+        args[xydims[1]] = ix
+        out = da.isel(**args)
+
+    if year_bnds:
+        year_bnds = np.asarray(year_bnds)
+        if len(year_bnds) == 1:
+            time_cond = da.time.dt.year == year_bnds
+        else:
+            time_cond = (da.time.dt.year >= year_bnds.min()) & (da.time.dt.year <= year_bnds.max())
```

Comment:
xarray supports slicing with strings, eg http://xarray.pydata.org/en/stable/indexing.html#indexing-with-dimension-names, which I feel could make your function more general. 

If for example you had start=None, end=None as parameters, you could do 
```
start = str(start) if start is not None
end = str(end) if end is not None
da.sel(time=slice(start, end))
```
This would let users get all the data from the start to a given year, or from a given year to the end.

Reply:
ok will change for both.. bbox and gridpoint ..  
FYI I used slicing for my initial bbox function but discovered that some .nc (nrcan I thin) have inverted latitudes (da,lat values start high and diminish... ). The raw datagrid is upside down if you use a simple imshow with no coords. To use slicing with this you need to invert the start / end values (e.g. slice(max(lat),min(lat)).   Can still work but requires checks to see if dims are ascending or descending ... ds.where work in either case..

Reply:
Hum, good point. Do you think this is something we could fix upstream ?

Reply:
Have added individual start and end year inputs .... have kept the ds.where logic for now vs slicing

Reply:
> Hum, good point. Do you think this is something we could fix upstream ?
We could probably apply a xarray.DataArray.sortby before slicing. http://xarray.pydata.org/en/stable/generated/xarray.DataArray.sortby.html
Would be curious about whether this ends up being slower than my current ds.where operation..